Google Map: Optimize dependencies

adamwoodnz opened the pull request:
Fixes #537 
Related: https://github.com/WordPress/wordcamp.org/issues/1128

Moves lighthouse perf score from ~68 to ~82 on localhost for the Events landing page.

Removes dependency on WordPress Components package and shifts lodash dependencies to individual packages (throttle and debounce).

### Spinner

#### Before

![Screenshot 2023-12-08 at 10 08 19 AM](https://github.com/WordPress/wporg-mu-plugins/assets/1017872/a5e819a2-710e-4de3-b898-319535ce5dcc)

#### After

![events spinner](https://github.com/WordPress/wporg-mu-plugins/assets/1017872/7f263e56-c271-445b-a16f-9e79b90d5dc1) |

### Scripts

#### Before (All (3.6 MiB), Unused Bytes (1.8 MiB))

![Screenshot 2023-12-08 at 9 56 10 AM](https://github.com/WordPress/wporg-mu-plugins/assets/1017872/d6663fb2-1a1b-4185-b17f-cd963df37e64)

#### After (All (1.9 MiB), Unused Bytes (810.6 KiB))

![Screenshot 2023-12-08 at 9 59 14 AM](https://github.com/WordPress/wporg-mu-plugins/assets/1017872/b5b9b0f7-dc95-4d77-be00-280f02b793bd)

### Testing

Should be tested with a site using the Screenshot Preview block, as it has dependency changes (Showcase?), and Events for the Google Map block.

adamwoodnz committed:
Replace WordPress components Spinner with local component

iandunn commented on mu-plugins/blocks/google-map/postcss/style.pcss:
```diff
@@ -74,8 +74,47 @@
+	100% {
+		transform: rotate(360deg);
+	}
+}
```
There's a lot of code we have to fork for this, it may be simpler to use the older spinner instead:

`<img src="<?php echo esc_url( includes_url( 'images/spinner-2x.gif' ) ); ?>" width="20" height="20" />`

We wouldn't need to create a JS component for it or anything, we could just put it in the PHP render:

https://github.com/WordPress/wporg-mu-plugins/blob/8c447146a557982cd78ce8695aa8ec04be9319db/mu-plugins/blocks/google-map/index.php#L91

The CSS would be smaller, and may be able to be pared down even further:

https://github.com/WordPress/WordPress/blob/46406afcc4cc938f44115c49ae775bad52039710/wp-admin/css/common.css#L2315-L2327

I don't feel strongly, though. Forking the Gutenberg spinner is fine if you prefer that.

adamwoodnz replied:
Let me see what that looks like

adamwoodnz replied:
Tried this out and I think we should stick with the svg. It's one less request, and a little more CSS, but this won't make much difference really. There were a couple of problems with the gif:

1. If we put it directly in the php we can't center it, as the map container provides the height, and that hasn't loaded yet
2. If we render it in the react component we can center it, but it doesn't look good on blue as it's been rasterized on white

![Screenshot 2023-12-08 at 11 52 45 AM](https://github.com/WordPress/wporg-mu-plugins/assets/1017872/4fb6b68f-d93a-4308-8992-5525d1169f49)

iandunn replied:
👍🏻

iandunn reviewed (approved):
This WFM 👍🏻 

It's disappointing that Gutenberg makes us replicate all of this. What's the point of having a component library if we can't use it on the front end? That's not something we can fix in this PR though.

adamwoodnz committed:
Remove dependency on WP component button

adamwoodnz committed:
Optimize lodash dependencies

lodash should be a dependency, it is used for runtime
Switch to individual packages

StevenDufresne commented on mu-plugins/blocks/google-map/postcss/style.pcss:
```diff
@@ -74,8 +73,47 @@
+		position: absolute;
+		left: var(--local-position);
+		top: var(--local-position);
+		color: var(--wp-components-color-accent, var(--wp-admin-theme-color, #3858e9));
```
Where does `--wp-components-color-accent` come from? Is that reliable?

adamwoodnz replied:
Hmm good spotting, I'll check

adamwoodnz replied:
Turns out no. Set it to blueberry and made it configurable in https://github.com/WordPress/wporg-mu-plugins/pull/538/commits/6f063a3ee3fce3011750f640bfea591f02f12487

StevenDufresne commented:
I haven't personally tested this, but the code looks good. 👍

adamwoodnz committed:
Set color for spinner and make configurable

iandunn commented on package.json:
```diff
@@ -67,6 +66,8 @@
-		"locutus": "^2.0.16"
+		"locutus": "^2.0.16",
+		"lodash.debounce": "^4.0.8",
+		"lodash.throttle": "^4.1.1"
```
Does this cause it to bundle that in the `front.js` build, rather than using the version that ships w/ Core?

If so, I wonder if that's actually worse for holistic performance, since the Core script would already be cached locally in a lot of places. I'm guessing that wouldn't show up in Lighthouse scores b/c it's not realistic enough to detect it.

adamwoodnz replied:
Yeah it does cause it to be bundled. Sorry, I wasn't aware of the core dependency, and it wasn't clear to me looking at our imports, where the comments list it as an External dependency, and not a Wordpress one:

<img src="https://github.com/WordPress/wporg-mu-plugins/assets/1017872/0ec0b6c6-9d06-4e75-8d2c-ca59d0cc8801" width="300">

iandunn replied:
I opened #543 to test it out, but it seems like the approach here is probably better 👍🏻
